std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,294 @@ unittest
+                    debug ElementType previousFront = source[i].front();
+                source[i].popFront();
+                static if (isCopyable!ElementType && !hasAliasing!ElementType)
+                    debug if (!source[i].empty)
+                    {
+                        assert(previousFront == source[i].front ||
+                               comp(previousFront, source[i].front),
+                               "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                    }
+                break sw;
+            }
         }
-        assert(false);
+        _lastFrontIndex = frontIndex;
     }
 
-    ///
-    @property auto ref ElementType front()
+    static if (isBidirectional)
     {
-        assert(!empty);
-        // Assume _crt is correct
-        foreach (i, U; Rs)
+        @property auto ref back()
         {
-            if (i < _crt) continue;
-            assert(!_r[i].empty);
-            return _r[i].front;
+            if (_lastBackIndex == size_t.max)
+            {
+                this._lastBackIndex = backIndex; // lazy initialization
+            }
+            final switch (_lastBackIndex)
+            {
+                foreach (i, _; Rs)
+                {
+                case i:
+                    assert(!source[i].empty);
+                    return source[i].back;
+                }
+            }
+        }
+
+        private size_t backIndex()
+        {
+            size_t bestIndex = size_t.max; // indicate undefined
+            Unqual!ElementType bestElement;
+            foreach (i, _; Rs)
+            {
+                if (source[i].empty) { continue; }
+                if (bestIndex == size_t.max || // either this is the first or
+                    comp(bestElement, source[i].back))
+                {
+                    bestIndex = i;
+                    bestElement = source[i].back;
+                }
+            }
+            return bestIndex;
+        }
+
+        void popBack()
+        {
+            if (_lastBackIndex == size_t.max)
+            {
+                this._lastBackIndex = backIndex; // lazy initialization
+            }
+            sw: final switch (_lastBackIndex)
+            {
+                foreach (i, R; Rs)
+                {
+                case i:
+                    static if (isCopyable!ElementType && !hasAliasing!ElementType)
+                        debug ElementType previousBack = source[i].back();
+                    source[i].popBack();
+                    static if (isCopyable!ElementType && !hasAliasing!ElementType)
+                        debug if (!source[i].empty)
+                        {
+                            assert(previousBack == source[i].back ||
+                                   comp(source[i].front, previousBack),
+                                   "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                        }
+                    break sw;
```

Comment:
as an aside, the label is not necessary - you may just leave the break here and insert one right after the foreach loop